README.md
# ekma: templating for ES6
```sh
npm install ekma --save
```

There are lots of templating languages and view systems for JS. Handlebars, Jade, Nunjucks, and React's JSX are just a few, and there are more every day! All of them expose a domain-specific language (DSL) and introduce plenty of complexity when all you want is regular JavaScript.

The common untapped feature is ES6 template strings, which ekma aims to fix in a portable and flexible way. ekma works well whether you're running native ES6 or transpiling to ES5 via Babel, Traceur, or TypeScript!

ekma includes TypeScript typings out of the box.

## Requirements
ekma requires an ES5 environment. That's it!

## Examples
ekma templates are just functions that return string-like objects!

Use the `$html` template tag to automatically escape inputs to fit into HTML:

```js
import { $html, $each, $if } from "ekma";

const myCollection = [5, 6, 7];

const template = () => $html`
	${ $if(myCollection.length === 0, () => `
		Nothing to show in the collection.
	`) }

	${ $each(myCollection, v => `
		Value: ${ v }
	`) }
`;

console.log(template().toString());
```

The return value of the template is an object with a `toString` method and the property `safeInHTML` set to `true`.

## API

This API is using TypeScript syntax for expressing types. For additional details, check [the source](https://github.com/LPGhatguy/ekma/blob/master/src/index.ts) of the module.

In these annotations, `Stringable` is an object that has a `toString` method, or is an actual string.

- `escapeHTML(html: string)`
	- escapes the given string as HTML
- `$html`
	- A template string tag that marks the string as HTML
	- Use `$${ value }` to output unescaped values in substitutions
	- Nested strings will automatically have correct escaping
- `$times(count: number, method: Function | Stringable)`
	- Outputs the given `Function` or `Stringable` `count` times.
- `$map(collection: Array, callback: Function | Stringable)`
	- Equivalent to `Array.prototype.map`, but evaluates functions
- `$each(collection: Array, callback: Function | Stringable)`
	- Iterates over the collection and joins the resulting array into a string
- `$alias(object: any, method: Function)`
	- Syntax sugar for method(object), used to shorten access chains
	- Similar to old JavaScript `with`
- `$if(condition: any, pass: Function | Stringable, fail: Function | Stringable)`
	- An `if` statement in function/expression form
	- Returns the result of `pass` if the condition is `true`
	- Otherwise returns the result of `fail`
- `$join(collection: Stringable[])`
	- Joins a collection of elements that may be safe in HTML

## Building
Building requires Node.js 5 or newer.

To build ekma from sources, run `npm run build` in the root of the project.

The main field specified in `package.json` is only valid after the library has been built, as compiled JS isn't checked into source control to reduce commit noise.

## License
ekma is licensed under the MIT license. See [LICENSE.md](LICENSE.md) for more details.

CHANGES.md
# ekma Changes

## 0.3.1
- Fixed escaping of `$html` values in `$if`

## 0.3.0
- Added `$join` method for an automatic `safeInHTML` version of `Array#join`
- `$each` and `$times` now correctly handle `safeInHTML` values
	- This feature requires that all results are marked `safeInHTML`!
- Relaxed type declarations on all iterators, they now default to `any` instead of `{}`.

## 0.2.2
- Fixed error in `$html` with null values in substitutions

## 0.2.1
- Fixed typings issue and added documentation

## 0.2.0
- Template escaping is now enabled by default in `$html`
- Removed `escapeQuotes` method, using traditional escaping is safe for quotes
- `$html` now returns an object with `#toString()` instead of a string

## 0.1.0
- Initial release
- Sourced from internal templating library with low overhead
- Developed alongside [Armature](https://github.com/LPGhatguy/armature)

es5/index.js
"use strict";
/**
 * Turns HTML-sensitive characters into HTML entities
 * Escapes all of &><"'`
 */
function escapeHTML(str) {
    if (typeof str !== "string" || str.length === 0) {
        return "";
    }
    return str.replace(/&/g, '&amp;')
        .replace(/>/g, '&gt;')
        .replace(/</g, '&lt;')
        .replace(/"/g, '&quot;')
        .replace(/'/g, '&#39;')
        .replace(/`/g, '&#96;');
}
exports.escapeHTML = escapeHTML;
/**
 * Turns double quotes into HTML entities in the given text.
 */
function escapeQuotes(str) {
    if (typeof str !== "string" || str.length === 0) {
        return "";
    }
    return str.replace(/"/g, "&quot;");
}
exports.escapeQuotes = escapeQuotes;
/**
 * An HTML snippet that's escaped by default.
 * Use $${"data"} to escape an input.
 */
function $html(literalSections) {
    var substs = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        substs[_i - 1] = arguments[_i];
    }
    var raw = literalSections.raw;
    var result = "";
    substs.forEach(function (subst, i) {
        var lit = raw[i];
        if (Array.isArray(subst)) {
            subst = subst.join("");
        }
        if (lit.endsWith("$")) {
            lit = lit.slice(0, -1);
        }
        else {
            if (subst == null) {
                subst = "null";
            }
            else {
                subst = escapeHTML(subst.toString());
            }
        }
        result += lit;
        result += subst;
    });
    result += raw[raw.length - 1];
    return result;
}
exports.$html = $html;
/**
 * Executes the given string or function a given number of times.
 */
function $times(n, method) {
    var buffer = [];
    for (var i = 0; i < n; i += 1) {
        if (typeof method === "function") {
            buffer.push(method(i));
        }
        else {
            buffer.push(method);
        }
    }
    return buffer.join("");
}
exports.$times = $times;
function $map(collection, method) {
    var buffer = [];
    var len = collection.length;
    for (var i = 0; i < len; i++) {
        if (typeof method === "function") {
            buffer.push(method(collection[i], i).toString());
        }
        else {
            buffer.push(method.toString());
        }
    }
    return buffer;
}
exports.$map = $map;
/**
 * The Armature replacement for Circular's `each`
 * Automatically joins the given strings with nothing between them.
 * Use `$map` if you want an array.
 */
function $each(collection, method) {
    return $map(collection, method).join("");
}
exports.$each = $each;
/**
 * Calls the given function with the given object as a parameter.
 * Useful for aliasing a value to a shorter name in a template.
 */
function $alias(object, callback) {
    return callback(object);
}
exports.$alias = $alias;
/**
 * Executes the function or string only if the condition given is truthy.
 */
function $if(condition, method, other) {
    if (condition) {
        if (typeof method === "function") {
            return method();
        }
        else if (method) {
            return method;
        }
    }
    else {
        if (typeof other === "function") {
            return other();
        }
        else if (other) {
            return other;
        }
    }
    return "";
}
exports.$if = $if;

//# sourceMappingURL=index.js.map


build.conf.js
"use strict";

const pack = require("./package.json");

// Load TypeScript from our dependencies
try {
	var typescript = require("typescript");
} catch(e) {
	console.error("Couldn't load some dependencies; try running 'npm install'");
	process.exit(1);
}

// Compiler options on the server
// typescript goes to gulp-typescript
const server = {
	typescript: {
		module: "commonjs",
		target: "ES5",
		moduleResolution: "node",
		typescript: typescript,
		noEmitOnError: true,
		experimentalDecorators: true,
		noLib: true
	}
};

const config = {
	// Omit this to turn off Browsersync
	// Passed to Browsersync
	browsersync: false,

	// Default preset
	preset: "default",

	// Collection of presets
	presets: {
		default: {
			watch: true,
			sourcemaps: true,
			minify: false,
			out: "es5"
		}
	},

	// List of pipelines to build and use
	pipelines: [
		// Build primary server files (node_modules/app/server/)
		{
			name: "Library (CommonJS)",
			config: server,
			type: "server",

			input: "src/**/*.ts",
			output: "",

			extraEntries: ["node_modules/typescript/lib/lib.es6.d.ts"],

			typingsOutput: "",
			typingsOutputType: "module",
			moduleName: pack.name,
			moduleEntryPoint: "index.ts"
		}
	]
};

module.exports = config;

debug/index.d.ts
interface Stringable {
    toString(): string;
}
export { Stringable };
export declare type EachIterator<T> = (value: T, key: number) => Stringable;
export declare type StringFunction = () => string;
/**
 * Turns HTML-sensitive characters into HTML entities
 * Escapes all of &><"'`
 */
export declare function escapeHTML(str: string): string;
/**
 * Turns double quotes into HTML entities in the given text.
 */
export declare function escapeQuotes(str: string): string;
/**
 * An HTML snippet that's escaped by default.
 * Use $${"data"} to escape an input.
 */
export declare function $html(literalSections: any, ...substs: any[]): string;
/**
 * Executes the given string or function a given number of times.
 */
export declare function $times(n: number, method: Function | string): string;
export declare function $map(collection: any[], method: EachIterator<any> | string): string[];
export declare function $map<T extends {}>(collection: T[], method: EachIterator<T> | string): string[];
/**
 * The Armature replacement for Circular's `each`
 * Automatically joins the given strings with nothing between them.
 * Use `$map` if you want an array.
 */
export declare function $each<T extends {}>(collection: T[], method: EachIterator<T> | string): string;
/**
 * Calls the given function with the given object as a parameter.
 * Useful for aliasing a value to a shorter name in a template.
 */
export declare function $alias<T extends {}>(object: T, callback: (v: T) => string): string;
/**
 * Executes the function or string only if the condition given is truthy.
 */
export declare function $if(condition: any, method: StringFunction | string, other?: StringFunction | string): string;


src/index.ts
interface Stringable {
	toString(): string;
	safeInHTML?: boolean;
}

export { Stringable };

export type EachIterator<T> = (value: T, key: number) => Stringable;
export type StringFunction = () => Stringable;

/**
 * Stopgap for checking of function args with TypeScript
 * Remove when TypeScript 2.0 is released.
 */
function isFunction(object: any): object is Function {
	return (typeof object === "function");
}

/**
 * Turns HTML-sensitive characters into HTML entities
 * Escapes all of &><"'`
 */
export function escapeHTML(str: string) {
	if (typeof str !== "string" || str.length === 0) {
		return "";
	}

	return str.replace(/&/g, '&amp;')
		.replace(/>/g, '&gt;')
		.replace(/</g, '&lt;')
		.replace(/"/g, '&quot;')
		.replace(/'/g, '&#39;')
		.replace(/`/g, '&#96;');
}

/**
 * An HTML snippet that's escaped by default.
 * Embedded instances of $html will not be extra-escaped.
 * Use $${"data"} to skip escaping of an input.
 */
export function $html(literalSections, ...substs): Stringable {
	const raw = literalSections.raw;
	let result = "";

	substs.forEach((subst, i) => {
		let lit = raw[i];

		if (Array.isArray(subst)) {
			subst = subst.join("");
		}

		if (lit.endsWith("$")) {
			lit = lit.slice(0, -1);
		} else {
			if (subst == null) {
				subst = "null";
			} else if (subst.safeInHTML) {
				subst = subst.toString();
			} else {
				subst = escapeHTML(subst.toString());
			}
		}

		result += lit;
		result += subst;
	});

	result += raw[raw.length - 1];

	return {
		toString() {
			return result;
		},
		safeInHTML: true
	};
}

/**
 * Executes the given string or function a given number of times.
 */
export function $times(count: number, method: Function | Stringable) {
	const buffer = [];

	for (let i = 0; i < count; i += 1) {
		if (isFunction(method)) {
			buffer.push(method(i));
		} else {
			buffer.push(method);
		}
	}

	return $join(buffer);
}

/**
 * Maps a list of values using the given function or string.
 */
export function $map<T>(collection: T[], method: EachIterator<T> | Stringable): Stringable[] {
	let buffer = [];
	let len = collection.length;

	for (let i = 0; i < len; i++) {
		if (isFunction(method)) {
			buffer.push(method(collection[i], i));
		} else {
			buffer.push(method);
		}
	}

	return buffer;
}

/**
 * Maps all values of the collection using $map, then joins the array.
 * Use `$map` to get an array.
 */
export function $each<T>(collection: T[], method: EachIterator<T> | Stringable): Stringable {
	const values = $map<T>(collection, method);

	return $join(values);
}

/**
 * Joins together a list of values and attempts to preserve safeInHTML if all values are.
 */
export function $join<T>(collection: T[]): Stringable {
	let isMixed = false;
	let isSafe: boolean;

	for (const value of collection) {
		const valueSafe = (value != null) && value["safeInHTML"];

		if (isSafe == null) {
			isSafe = valueSafe;
		} else {
			if (isSafe !== valueSafe) {
				isMixed = true;
			}

			isSafe = isSafe && valueSafe;
		}
	}

	const str = collection.join("");

	if (isMixed) {
		console.error("Mixed safe/unsafe content found in an ekma loop. This should be avoided as of ekma 0.3.0!");
	}

	if (isSafe) {
		return {
			toString() {
				return str;
			},
			safeInHTML: true
		};
	} else {
		return str;
	}
}

/**
 * Calls the given function with the given object as a parameter.
 * Useful for aliasing a value to a shorter name in a template.
 */
export function $alias<T>(object: T, method: (v: T) => Stringable): Stringable {
	return method(object);
}

/**
 * Executes the function or string only if the condition given is truthy.
 */
export function $if(condition: any, method: StringFunction | Stringable, other?: StringFunction | Stringable): Stringable {
	if (condition) {
		if (isFunction(method)) {
			return method();
		} else if (method) {
			return method;
		}
	} else {
		if (isFunction(other)) {
			return other();
		} else if (other) {
			return other;
		}
	}

	return "";
}